packages/remark-images-download/src/index.js

```diff
@@ -2,212 +2,330 @@ const fileType = require('file-type')
-                if (!isImage(headers)) {
-                  reject(new Error(`Content-Type of ${url} is not an image/ type`))
-                }
-
-                const fileSize = getSize(headers)
-                if (maxFileSize && fileSize > maxFileSize) {
-                  reject(new Error(
-                    `File at ${url} weighs ${headers['content-length']}, ` +
-                    `max size is ${maxFileSize}`))
-                }
-
-                if (dirSizeLimit && (totalDownloadedSize + fileSize) >= dirSizeLimit) {
-                  reject(new Error(
-                    `Cannot download ${url} because destination directory reached size limit`))
-                }
-
-                totalDownloadedSize += fileSize
-              })
-              .on('response', (res) => {
-                res.once('data', (chunk) => {
-                  const type = fileType(chunk) || {mime: ''}
-                  if (type.mime.slice(0, 6) !== 'image/' && !isSvg(chunk.toString())) {
-                    if (type.mime) {
-                      reject(new Error(`Mime of ${url} not allowed: '${type.mime}'`))
-                    } else {
-                      reject(new Error(`Could not detect ${url} mime type, not SVG either`))
-                    }
-                  }
-                })
-              })
-              .on('error', (err) => {
-                reject(err)
-              })
-              .pipe(writeStream(destination))
-          }).catch((err) => {
-            vfile.message(err, position, url)
-          })
-
-          promises.push({offset: position.offset, promise: promise})
-        })
+      const options = Object.assign(
+        {},
+        parsedUrl,
+        {timeout: httpRequestTimeout}
+      )
+
+      const req = proto.get(options, res => {
+        const {headers, statusCode} = res
+        let error
 
-        if (!promises.length) {
-          return new Promise((resolve, reject) =>
-            rimraf(destinationPath, (err) => {
-              if (err) reject(err)
-              resolve()
-            }))
+        const fileSize = getSize(headers)
+
+        if (statusCode !== 200) {
+          error = new Error(`Received HTTP${statusCode} for: ${url}`)
+        } else if (!isImage(headers)) {
+          error = new Error(`Content-Type of ${url} is not an image/ type`)
+        } else if (maxFileSize && fileSize > maxFileSize) {
+          error = new Error(
+            `File at ${url} weighs ${headers['content-length']}, ` +
+            `max size is ${maxFileSize}`
+          )
+        }
+
+        if (error) {
+          req.abort()
+          res.resume()
+          reject(error)
+          return
         }
 
-        vfile.data.imageDir = destinationPath
-
-        // Use offsets to ensure execution order
-        // we don't want to download them in (possibly) random order.
-        // More importantly: this makes tests stable.
-        promises.sort((a, b) => b.offset - a.offset)
```

Comment:
I removed this sorting because it seems the `visit` functions always walks the AST in the same order, so “tasks” and `vfile.messages` are always ordered according to the markdown source.
Moreover, files are downloaded and copied in parallel.

Reply:
IIRC there was a test for the max directory size that would try downloading 4 images of size x in a directory with a size limit slightly under 4x.

Since the downloads were parallel and since the test was also checking that only 3 out of the 4 urls had been replaced before reaching the size limit, this hack was there to make this test deterministic.

It was probably a bad fix for a bad idea anyway. :)